deluge/ui/gtkui/gtkui.py

```diff
@@ -239,35 +240,51 @@ def nsapp_open_file(osxapp, filename):
 
         # Twisted catches signals to terminate, so have it call a pre_shutdown method.
         reactor.addSystemEventTrigger("before", "shutdown", self.pre_shutdown)
+        reactor.addSystemEventTrigger("before", "gtkui_close", self.close)
+
         reactor.callWhenRunning(self._on_reactor_start)
+        self.closing = False
+
+        def gtkui_sigint_handler(num, frame):
+            log.debug("SIGINT signal caught - firing event: 'gtkui_close'")
+            reactor.callLater(0, reactor.fireSystemEvent, 'gtkui_close')
+
+        signal.signal(signal.SIGINT, gtkui_sigint_handler)
 
         # Initialize gdk threading
         gtk.gdk.threads_enter()
         reactor.run()
-        self.shutdown()
-        gtk.gdk.threads_leave()
-
-    def shutdown(self, *args, **kwargs):
-        log.debug("gtkui shutting down..")
-
-        component.stop()
 
         # Process any pending gtk events since the mainloop has been quit
         while gtk.events_pending():
-            gtk.main_iteration(0)
+            gtk.main_iteration_do(block=False)
+            time.sleep(0.0001)
```

Comment:
They do explain the purpose of the changes:

- `Fix #535 fix saving column widths`
- `Fix exception when quitting in classic mode`
- `Fix saving torrent state on fresh configs`

So if all these work now it should be ok to remove the code. I haven't tested if the column widths are saved. State should be properly saved, but ideally there would be unit tests for this.

When testing with print statements the save state function did run properly, because the whole component shutdown process is now ensured to be performed before the reactor is shutting down.

I have not tested on Windows, which probably would be a good idea.